app/controllers/users/nbp_wallet_controller.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+require 'rqrcode'
+
+module Users
+  class NbpWalletController < ApplicationController
+    skip_before_action :require_user!
+    skip_after_action :verify_authorized
+    before_action :set_provider_details
+    before_action :set_connector_api
+
+    def connect
+      truncated_reference = @connector.create_relationship_template(@provider_uid)
+
+      @template_url = "nmshd://tr##{truncated_reference}"
+      @truncated_reference = truncated_reference
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      redirect_to new_user_registration_path, alert: t('common.errors.generic_try_later')
+    end
+
+    def qr_code
+      send_data RQRCode::QRCode.new("nmshd://tr##{params[:truncated_reference]}").as_png(border_modules: 0), type: 'image/png'
+    end
+
+    def relationship_status
+      relationship = @connector.pending_relationship_for_nbp_uid(@provider_uid)
+
+      if relationship.present?
+        render json: {status: 'ready'}
+      else
+        render json: {status: 'waiting'}
+      end
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      redirect_to nbp_wallet_connect_users_path, alert: t('common.errors.generic')
+    end
+
+    def finalize
+      @relationship = @connector.pending_relationship_for_nbp_uid(@provider_uid)
+      abort_and_refresh and return if @relationship.blank?
+
+      accept_and_create_user
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      abort_and_refresh
+    end
+
+    private
+
+    def accept_and_create_user # rubocop:disable Metrics/AbcSize
+      relationship_changes = @relationship[:changes]
+      abort_and_refresh if relationship_changes.size != 1
+
+      user_attributes = @connector.parse_userdata_in_rel_change(relationship_changes.first)
+
+      @user = User.new_from_omniauth('nbp', @provider_uid, user_attributes)
+      @user.skip_confirmation_notification!
+      return abort_and_refresh("Could not create User: #{@user.errors.full_messages.join(', ')}") unless @user.save
```

Comment:
I discovered another edge case here: When entering just a tab (or space) character in one of the fields, the attribute is not empty so that it can be submitted in the app. However, this value is not considered to be present in Rails (`" ".present? => false`), resulting in a call to `abort_and_refresh`. Besides rejecting (what was done successfully) and noting, I assume we cannot do much about it? Unfortunately, when the contact request was rejected, it was still showing in my app, preventing me from using the same profile again for a next try (see https://github.com/openHPI/codeharbor/pull/1326#issuecomment-2120203235).

Reply:
I tested it but could not reproduce the problem. The user was rejected and an error message shown correctly but I was still able to scan a new QR code and register successfully (The app does display the failed attempts as "pending Relationship" though).

Reply:
Alright. I tested again and the issue in my version of the app was that it did not refresh the status automatically. Instead, opening the link again presented me with the contact details (detailling that we connected successfully today). I had to quit the app completely and re-open it to be able to scan another QR code. Then, it allowed me to re-enter my user details (with a modified last name) and the contact request was successful.

https://github.com/openHPI/codeharbor/assets/7300329/fc12ec59-0d97-476a-ad97-a36f2cd22829

As you mentioned already, the rejected request is still shown as "pending" in the app, but this is nothing we can fix (the server and connector logs clearly show that we rejected the relationship).

<img width="1136" alt="Bildschirmfoto 2024-05-26 um 14 45 19" src="https://github.com/openHPI/codeharbor/assets/7300329/d0c6e806-72af-40fe-a1ca-6c2a19617ec7">

That having said, I think we can resolve this conversation and move the bug to our follow-up bug reporting document in Nextcloud.